### ISSUE TITLE
Subtly incorrect import resolution with relative imports

### ISSUE THREAD

**Issue by smackesey** [microsoft/pyright#2140]
**Describe the bug**

PyRight incorrectly resolves imports in certain special circumstances, when an imported value uses the same name as a submodule.

**To reproduce**

Create the following test structure:

```python
### a/__init__.py
import a.ambiguous
from .b import ambiguous
from .ambiguous import foo

### a/b.py
ambiguous = 'string'

### a/ambiguous.py
foo = None

### test.py
from a import ambiguous
print(ambiguous)
```

Now make a hover request over the `ambiguous` imported in `test.py`. You will see it is reported as being of type module, i.e. it is resolving to the module defined by `a/ambiguous.py` instead of the value set in `a/b.py`. This is incorrect, which you can confirm by running `test.py`, which will print `string`.

**VS Code extension or command-line**
I'm using coc-pyright in neovim.

**Additional context**
I don't fully grasp the Python import system, but as best I can tell here is what's going on:

- The statement `from .ambiguous import foo` has the side effect of setting `ambiguous` as a local variable IF AND ONLY IF the module that `.ambiguous` (in this case, `a.ambiguous`) resolves to has not yet been loaded (why Python, why...?).
- In the above code, since we import `a.ambiguous` on the first line, `a.ambiguous` has already been loaded, so `from .ambiguous import foo` does not set the variable `ambiguous`. Therefore, ambiguous retains the value set on the preceding line (i.e. the string value defined in and imported from `a.b`).
- You can confirm the above by toggling a comment on the `import a.ambiguous` line and running `test.py`. The output of is dependent on whether this is line is commented. If it's commented, then the output is a string representation of module `a.ambiguous` (matching PyRight's interpretation). When not commented the output is `string`.
- PyRight doesn't take account of Python's weird behavior here, and so it ends up incorrectly typing variables.

**NOTE**: This contrived example comes out of my attempt to replicate weird behavior I found using PyRight with the popular data science library [Dagster](https://github.com/dagster-io/dagster).

**Comment by erictraut** [comment]
Yes, the Python import mechanism leaves much to be desired. The runtime logic for import has order-dependent side effects that are difficult or impossible to model statically (and difficult for any programmer to reason about!). Python code that relies on such side effects is arguably buggy, or at least extremely fragile and likely to break when seemingly innocuous changes are made in other source files or libraries.

In pyright, we don't attempt to model such side effects, since we'd end up being wrong as often as we're right. Instead, we always assume that all import-related effects across files are independent of each other (i.e. we assume that module imports may be resolved in any arbitrary order at runtime). And we assume that import-related effects _within a file_ are applied in file order with the last one "winning". That's why you're seeing in the example above that the type of `ambiguous` is resolved as a `module` rather than a `str`. If you swap the order of the last two lines in `a/__init__.py`, you'll see that the result changes accordingly.

One could argue that in this particular example, the behavior is deterministic and statically determinable, but this quickly falls apart with a small change to the code. For example, what is the correct behavior in this case?

```python
### a/__init__.py
if some_condition:
    import a.ambiguous
from .b import ambiguous
from .ambiguous import foo
```

Interestingly, it appears that mypy always determines the type to be `str` regardless of the ordering, which also does not match the runtime behavior. So mypy also falls back on a heuristic, but I haven't been able to determine what rules it uses.

My inclination is to stick with our current heuristic rather than attempting to do something more complicated, which will also be wrong some of the time — but harder to predict and reason about.

With the current simple heuristic in place, you can get the intended behavior simply by reordering the statements in the file. Of course, it would be even better to eliminate the dependency on import side effects regardless of statement ordering.

**Comment by smackesey** [comment]
Thanks for this detailed explanation. I can see the problem you illustrate when imports are dependent on some runtime condition, and I see that it's unfortunate that different static analyzers end up making different arbitrary assumptions, and that there's no good solution to this.

The issue for me is that I don't have control over the problem code in the library I'm using (Dagster). I think Dagster was initially typed with mypy in mind, which is probably why no one noticed the weirdness that I've run into. I'll let them know and maybe they can rearrange imports to get rid of this problem.

One thing that might be helpful on PyRight's end is a page in an FAQ or something explaining these and similar problems, so authors know exactly what patterns to avoid.

**Comment by hmc-cs-mdrissi** [comment]
Can this potential name shadowing be detected and warned on? That wouldn't fix the issue of knowing the correct type, but it'd be enough to help avoid code that causes it in the future.

Unsure if the warning makes more sense here vs pylint. Name shadowing feels more of a linter thing than a type checker thing besides the issue here there's inconsistent types across different names.

**Comment by smackesey** [comment]
I posted an issue in the Dagster repo that shows how I originally encountered this problem:

https://github.com/dagster-io/dagster/issues/4439

**Comment by smackesey** [comment]
OK, I've noticed something else. @erictraut you say:

> we assume that import-related effects within a file are applied in file order with the last one "winning"

However, this doesn't seem to be 100% correct. In my example, try changing the relative import `from .ambiguous import foo` to the absolute `from a.ambiguous import foo`. Now, from what I can tell, to Python itself it makes no difference, there is an "import-related effect" of defining `ambiguous` if and only if `a.ambiguous` has not already been loaded. But PyRight treats the two cases differently. It only accounts for the import-related effect if its a relative import, not an absolute.

**Comment by erictraut** [comment]
The statement `from a.ambiguous import foo` assigns a value to the local symbol `foo`, but it does not assign any value to the local symbol `ambiguous`. 

```python
from a.ambiguous import foo
print(ambiguous) # Runtime exception occurs because "ambiguous" is not bound
```

The statement `from .ambiguous import foo` assigns a value to the local symbol `foo` and (implicitly) assigns to the local symbol `ambiguous`. Relative imports have this implicit behavior.

```python
from .ambiguous import foo
print(ambiguous) # No exception because "ambiguous" is bound to a module
```

When pyright determines the type of the local symbol `ambiguous`, it looks at all the places in the file where a value is assigned to the symbol. If `ambiguous` is re-exported and then imported by another module, it assumes that the last assignment "wins".

For example:
```python
# a/__init__.py
from .ambiguous import foo
ambiguous = "some string"

# a/b.py
from a import ambiguous
reveal_type(ambiguous) # str
```

```python
# a/__init__.py
ambiguous = "some string"
from .ambiguous import foo

# a/b.py
from a import ambiguous
reveal_type(ambiguous) # module
```

**Comment by smackesey** [comment]
>The statement from a.ambiguous import foo assigns a value to the local symbol foo and a, but it does not assign any value to the local symbol ambiguous.

Are you sure? I believe the behavior is that both `from a.ambiguous import foo` *and* `from .ambiguous import foo` assign to the local symbol `ambiguous`, if and only if `a.ambiguous` has not already been loaded-- AFAICT it doesn't have anything to do with relative imports (despite the issue name, I am learning about the Python import system working through  this!).

That is because the first time a submodule is loaded, its name is bound in the parent package (this is described [here](http://python-notes.curiousefficiency.org/en/latest/python_concepts/import_traps.html#the-submodules-are-added-to-the-package-namespace-trap)).

This is also the behavior I'm seeing on my machine.

**Comment by erictraut** [comment]
You're confusing the namespace of module `a` with the local namespace of the source file that contains the import statement. This import may modify other files' namespaces as a side effect, but we don't attempt to model cross-file side effects for reasons I mentioned above.

**Comment by smackesey** [comment]
>You're confusing the namespace of module a with the local namespace of the source file that contains the import statement.

Maybe I'm confused, I don't think so. Do you know somewhere I can read about this relative-import rule?

For the example under discussion, the "namespace of module `a`" and the "local namespace of the source file containing the import statement" are the same, aren't they? This is generally true of one-level deep relative imports-- the namespace of the parent module and the local namespace of the importing source file are the same. So I think you're convinced that there is some special rule for relative imports, when in fact you're just seeing the effect of the rule I stated above-- that a name is bound in the parent module the first time the submodule is loaded.

Consider, if this rule were correct:

> The statement from .ambiguous import foo assigns a value to the local symbol foo and (implicitly) assigns to the local symbol ambiguous. Relative imports have this implicit behavior.

Then `from .ambiguous import foo` should assign to `ambiguous` regardless if `a.ambiguous` has already been loaded. But it doesn't. It only assigns to `ambiguous` if `a.ambiguous` *has not yet* been loaded. You can confirm:

```python
### a/__init__.py
import a.ambiguous
ambiguous = 5
from .ambiguous import foo  # this will not assign anything to `ambiguous`
print(ambiguous)  # => prints 5

### a/ambiguous.py
foo = None

### test.py
from a import ambiguous
```

Running test.py will give `5` because `a.ambiguous` was already loaded.

Similarly, this is incorrect:

>The statement from a.ambiguous import foo assigns a value to the local symbol foo and a, but it does not assign any value to the local symbol ambiguous.

I'm not sure of any conditions under which it assigns to local symbol `a`. Change `test.py` to this:

```python
from a.ambiguous import foo
print(a)  # => NameError
```

However, if we're inside `a/__init__.py`, then the statement behaves exactly equivalently to `from .ambiguous import a`. That is, it assigns to `ambiguous` if and only if `a.ambiguous` was not already loaded.

The upshot of my claims here are that `from <absolute-path-to-parent>.x import y` and `from .x import y` are semantically equivalent and therefore should be treated the same by PyRight, which they are currently not.

**Comment by erictraut** [comment]
Yes, my statement "Relative imports have this implicit behavior" applies only for `__init__.py` files. And pyright applies it only in this case. We added this special-case logic because some libraries rely on this ability.

```python
# a/__init__.py
ambiguous = 5
from .ambiguous import foo

print(ambiguous)

# Library assumes that "ambiguous" refers to a module. Ideally,
# code would make this explicit (`from . import ambiguous`) if
# that was their intent, but this is unfortunately a common pattern.
```

You're saying that this sequence has the same side effects as the one above and suggesting that we should also add special-case logic to handle it, similar to what we did in the previous case. In practice, we haven't see any major libraries rely on this side effect. I would be surprised if they did. Adding this special-case logic is more difficult than the previous one because it requires the analyzer to know that the current file `__init__.py` is the one referred to by `a` in this particular context.

```python
# a/__init__.py
ambiguous = 5
import a.ambiguous

print(ambiguous)

# This import has the side effect of populating the local "ambiguous"
# symbol because we're in the `a/__init__.py` file, but I wouldn't expect a
# library to rely on this. We haven't seen this behavior in practice.
```

The following sequence of statements is way more problematic in that it relies on order-dependent stateful side effects of the python import loader. I hope I've convinced you that a static type checker cannot and should not attempt to model these side effects — and that a library should not rely on them.

```python
### a/__init__.py
import a.ambiguous
ambiguous = 5
from .ambiguous import foo
print(ambiguous)
```

**Comment by smackesey** [comment]
Thanks for the reply. Yes, I agree that there is no need to try to model the side effects (i.e. try to figure out if the library was already loaded).

My core argument is simply that there are no conditions under which, inside an `__init__.py`, `from .sub import x` and `from <path-to-parent>.sub import x` differ, and therefore they should be treated the same. However, if the latter is more difficult to implement, I can see why it was not done.

BTW, this emerges out of the issue I opened at Dagster: https://github.com/dagster-io/dagster/issues/4439#issuecomment-892270435

The dev thought maybe changing to the absolute-import of a submodule would fix PyRight's incorrect typing of a value. I was surprised to see that it did, and I now understand that is bc PyRight does not bind a name in the local namespace using the absolute import form despite the semantic equivalence to the relative.

But I advised them that they should just avoid exporting values that collide with submodule names, since apparently this practice creates a time bomb bc of side effects.
